modules/gapi/test/internal/gapi_int_pattern_matching_test.cpp

```diff
@@ -994,4 +1002,293 @@ TEST(PatternMatching, adeSmokeTest)
+
+    compiled(input, output_transformed);
+
+    EXPECT_TRUE(AbsExact()(output_baseline, output_transformed));
+}
+
+TEST(PatternMatchingFull, SubstituteGraphInTheMiddle)
+{
+    cv::Size in_sz(640, 480);
+    cv::Mat input(in_sz, CV_8UC3);
+    cv::randu(input, cv::Scalar::all(0), cv::Scalar::all(100));
+    cv::Mat output_baseline, output_transformed;
+
+    cv::Size out_sz(100, 100);
+    double fx = 0.0, fy = 0.0;
+    int interpolation = cv::INTER_LINEAR;
+
+    const auto compile_args = [] () {
+        return cv::compile_args(cv::gapi::GKernelPackage{});
+    };
+
+    const auto ade_get_graph = [&] (const cv::GComputation& c) {
+        cv::gimpl::GCompiler compiler1(c, cv::descr_of(cv::gin(input)), compile_args());
+        auto ade_graph_ptr = compiler1.generateGraph();
+        compiler1.runPasses(*ade_graph_ptr);  // FIXME: only for dumpDot
+        return ade_graph_ptr;
+    };
+
+    // Main
+    std::unique_ptr<cv::GComputation> cmg;
+    GMat in;
+    GMat tmp = cv::gapi::bitwise_or(in, in);
+    GMat to_be_replaced = cv::gapi::resize(tmp, out_sz, fx, fy, interpolation);
+    GMat out = cv::gapi::mul(to_be_replaced, cv::gapi::bitwise_not(to_be_replaced));
+    cv::GComputation(cv::GIn(in), cv::GOut(out)).apply(input, output_baseline);
+    cmg.reset(new cv::GComputation(cv::GIn(in), cv::GOut(out)));
+    auto ade_mg = ade_get_graph(*cmg);
+
+    matching_test::myDumpDotToFile(*ade_mg, "main.dot");
+
+    // Pattern
+    std::unique_ptr<cv::GComputation> cpg;
+    {
+        GMat in;
+        GMat out = cv::gapi::resize(in, out_sz, fx, fy, interpolation);
+        cpg.reset(new cv::GComputation(cv::GIn(in), cv::GOut(out)));
+    }
+
+    auto ade_pg = ade_get_graph(*cpg);
+    matching_test::myDumpDotToFile(*ade_pg, "pattern.dot");
+
+    // Substitute
+    std::unique_ptr<cv::GComputation> csg;
+    {
+        GMat in;
+        GMat b, g, r;
+        std::tie(b, g, r) = cv::gapi::split3(in);
+        auto resize = std::bind(&cv::gapi::resize, std::placeholders::_1, out_sz, fx, fy,
+            interpolation);
+        GMat out = cv::gapi::merge3(resize(b), resize(g), resize(r));
+        csg.reset(new cv::GComputation(cv::GIn(in), cv::GOut(out)));
+    }
+
+    auto ade_sg = ade_get_graph(*csg);
+    matching_test::myDumpDotToFile(*ade_sg, "substitute.dot");
+
+    // Substitute && run
+    cv::gimpl::GCompiler compiler(*cmg, cv::descr_of(cv::gin(input)), compile_args());
+    ASSERT_TRUE(compiler.transform(*ade_mg, *cpg, *csg));
+
+    compiler.runPasses(*ade_mg);
+    compiler.compileIslands(*ade_mg);
+    auto compiled = compiler.produceCompiled(std::move(ade_mg));
+
+    compiled(input, output_transformed);
+
+    ASSERT_TRUE(AbsExact()(output_baseline, output_transformed));
+}
+
+using CompCreator = std::function<cv::GComputation()>;
+struct PatternMatchingHighLevel :
+    TestWithParam<std::tuple<CompCreator, CompCreator, CompCreator>> {};
+
+TEST_P(PatternMatchingHighLevel, UnifiedTest) {
```

Comment:
Comment to reviewers: please look at this test first to understand the big picture